notebook/static/edit/js/editor.js

```diff
@@ -200,32 +212,111 @@ function(
             function (model) {
                 that.file_path = model.path;
                 that.events.trigger('file_renamed.Editor', model);
+                that.last_modified = new Date(model.last_modified);
                 that._set_mode_for_model(model);
                 that._clean_state();
             }
         );
     };
     
-    Editor.prototype.save = function () {
+
+    /**
+     * Save this file on the server.
+     *
+     * @param {boolean} check_last_modified - checks if file has been modified on disk
+     * @return {Promise} - promise that resolves when the notebook is saved.
+     */
+    Editor.prototype.save = function (check_last_modified) {
         /** save the file */
         if (!this.save_enabled) {
             console.log("Not saving, save disabled");
             return;
         }
+
+        // used to check for last modified saves
+        if (check_last_modified === undefined) {
+            check_last_modified = true;
+        }
+
         var model = {
             path: this.file_path,
             type: 'file',
             format: 'text',
             content: this.codemirror.getValue(),
         };
         var that = this;
-        // record change generation for isClean
-        this.generation = this.codemirror.changeGeneration();
-        that.events.trigger("file_saving.Editor");
-        return this.contents.save(this.file_path, model).then(function(data) {
-            that.events.trigger("file_saved.Editor", data);
-            that._clean_state();
-        });
+
+        var _save = function () {
+            that.events.trigger("file_saving.Editor");
+            return that.contents.save(that.file_path, model).then(function(data) {
+                // record change generation for isClean
+                that.generation = that.codemirror.changeGeneration();
+                that.events.trigger("file_saved.Editor", data);
+                that.last_modified = new Date(data.last_modified);
+                that._clean_state();
+            });
+        };
+
+        /* 
+         * Gets the current working file, and checks if the file has been modified on disk. If so, it
+         * creates & opens a modal that issues the user a warning and prompts them to overwrite the file.
+         * 
+         * If it can't get the working file, it builds a new file and saves.
+         */ 
+        if (check_last_modified) {
+            return this.contents.get(that.file_path, {content: false}).then(
+                function check_if_modified(data) {
+                    var last_modified = new Date(data.last_modified);
+                    // We want to check last_modified (disk) > that.last_modified (our last save)
+                    // In some cases the filesystem reports an inconsistent time,
+                    // so we allow 0.5 seconds difference before complaining.
+                    if ((last_modified.getTime() - that.last_modified.getTime()) > 500) {  // 500 ms
```

Comment:
We recently merged a pull request to make this 0.5s difference configurable in the notebook - #3273 . Could you have a look at that and copy the functionality here? Thanks!